src/gridfs/upload.rs

```diff
@@ -173,3 +179,287 @@ impl GridFsBucket {
+            _ => Poll::Ready(Ok(())),
+        }
+    }
+
+    fn poll_close(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<tokio::io::Result<()>> {
+        let stream = self.get_mut();
+
+        let future = match &mut stream.state {
+            State::Idle(buffer) => {
+                let buffer = buffer.take().unwrap();
+
+                let file = FilesCollectionDocument {
+                    id: stream.id.clone(),
+                    length: stream.current_n as u64 * stream.chunk_size as u64
+                        + buffer.len() as u64,
+                    chunk_size: stream.chunk_size,
+                    upload_date: DateTime::now(),
+                    filename: stream.filename.take(),
+                    metadata: stream.metadata.take().unwrap(),
+                };
+
+                let new_future = close(stream.bucket.clone(), buffer, file).boxed();
+                stream.state.set_closing(new_future)
+            }
+            // This case is effectively unreachable, as the AsyncWriteExt methods take &mut self and
+            // poll the futures to completion. If a user were to call these polling methods directly
+            // and intersperse futures, we should just pend here until the writing future resolves.
+            State::Writing(_) => return Poll::Pending,
+            State::Closing(future) => future,
+            State::Closed => return Poll::Ready(Err(get_closed_error())),
+        };
+
+        let result = match future.poll_unpin(cx) {
+            Poll::Ready(result) => result,
+            Poll::Pending => return Poll::Pending,
+        };
+
+        stream.state = State::Closed;
+        match result {
+            Ok(()) => Poll::Ready(Ok(())),
+            Err(error) => Poll::Ready(Err(error.into_futures_io_error())),
+        }
+    }
+}
+
+async fn write_bytes(
+    bucket: GridFsBucket,
+    mut buffer: Vec<u8>,
+    starting_n: u32,
+    chunk_size: u32,
+    files_id: Bson,
+) -> Result<(u32, Vec<u8>)> {
+    bucket.create_indexes().await?;
+
+    let mut n = 0;
+    let mut chunks = vec![];
+
+    while buffer.len() as u32 - (n * chunk_size) >= chunk_size {
+        let start = n * chunk_size;
+        let end = (n + 1) * chunk_size;
+        let chunk = Chunk {
+            id: ObjectId::new(),
+            files_id: files_id.clone(),
+            n: starting_n + n,
+            data: RawBinaryRef {
+                subtype: BinarySubtype::Generic,
+                bytes: &buffer[(start as usize)..(end as usize)],
+            },
+        };
+        n += 1;
+        chunks.push(chunk);
+    }
+
+    bucket.chunks().insert_many(chunks, None).await?;
+    buffer.drain(..(n * chunk_size) as usize);
+
+    Ok((n, buffer))
+}
+
+async fn close(bucket: GridFsBucket, buffer: Vec<u8>, file: FilesCollectionDocument) -> Result<()> {
+    let insert_result: Result<()> = async {
+        if !buffer.is_empty() {
+            let final_chunk = Chunk {
```

Comment:
might be worth adding a debug assert here that `buf.len() < chunk_size` for testing purposes.